### PR TITLE
Issue 8 reorganize

### LINKED ISSUE #8: New project API
Reorganize the project structure and API.

From the API point of view, all the functionalities will be implemented as methods of a single class, called SteganoGAN.
An additional class called ImageLoader will be created, responsible of loading images from a folder.

The class will be `steganogan.steganogan.SteganoGAN`, which will also be importable as `steganogan.SteganoGAN`.

This class will receive as input:

* data_depth
* hidden_size
* encoder: Class to use as the Encoder
* decoder: Class to use as the Decoder
* critic: Class to use as the Critic
* models_path: path to the folder where models are serialized (optional)

And it will implement these methods:

* `fit(training, validation)`: Will load a dataset and fit new models. It will receive two arguments, `training` and `validation`, which will be instances of the ImageLoader class.
* `encode(message, image, output)`: Will encode the `message` using the `image` and store it as `output`.
* `decode(image)`: Will decode the message found in the given `image`.
* `save(path)`: Will save the fitted model in the given path. It will raise an exception if the model is not fitted.
* `load(path)`: Classmethod. Will load an instance of the SteganoGAN from the given path.

### PR DESCRIPTION
**Pull request by csala**
Resolve #8 

* Move fit logic to SteganoGAN class
* Major refactoring to improve code readability and efficiency
* Add notebooks and pretrained basic and dense models

### CODE REVIEW

File: steganogan/models.py [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+            generated = self.encoder(cover, payload)
+            cover_score = self._critic(cover)
+            generated_score = self._critic(generated)
+
+            self.critic_optimizer.zero_grad()
+            (cover_score - generated_score).backward(retain_graph=True)
+            self.critic_optimizer.step()
+
+            for p in self.critic.parameters():
+                p.data.clamp_(-0.1, 0.1)
```
[the remark this answers is not in the document]
Comment (csala) [reply steganogan/models.py]: Let's keep that for a later PR.

File: steganogan/models.py [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+
+        return generated, payload, decoded
+
+    def _critic(self, image):
+        """Evaluate the image using the critic"""
+        return torch.mean(self.critic(image))
+
+    def _get_optimizers(self):
+        _dec_list = list(self.decoder.parameters()) + list(self.encoder.parameters())
+        critic_optimizer = Adam(self.critic.parameters(), lr=1e-4)
```
[the remark this answers is not in the document]
Comment (csala) [reply steganogan/models.py]: Let's keep that for a later PR.

File: steganogan/models.py [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+        """Generate random data ready to be hidden inside the cover image.
+
+        Args:
+            cover (image): Image to use as cover.
+
+        Returns:
+            generated (image): Image generated with the encoded message.
+        """
+        N, _, H, W = cover.size()
+        cover.to(self.device)
```
[the remark this answers is not in the document]
Comment (csala) [reply steganogan/models.py]: This is a "private" method, that should be called only from inside the fit method, so it shouldn't be a problem.
However, right now when we reach this point the image is already set to the right device, so this like is just unnecessary and will be removed.

File: README.md [hunk 4]
```diff
@@ -1,43 +1,81 @@
...
+```
 
-model = Steganographer()
-model.encode(in_path, "Hello!", out_path)
+For development, you can use the `make install-develop` target instead in order to install all
+the required dependencies for testing, code linting and notebook running.
+
+# Usage
+
+The simplest way to use **SteganoGAN** is loading one of the pretrained models  which have been
```
[the remark this answers is not in the document]
Comment (k15z) [reply README.md]: Perfect!

**Pull request merged**